### PR TITLE
Convert stats and diagnostics to new API

### PR DESCRIPTION
**Pull request by ColCarroll**
These changes get the test suite for `diagnostics.py` to pass, with everything being xarray based. Now will need to work on `stats.py`, which will remove (most of) the pandas requirements.  I am working on that, but will probably need help for some of the functions (#231) .

Note that most functions in `diagnostics.py` work with either a dataset, or a numpy array.

### CODE REVIEW

**Comment by ColCarroll** [comment]
I _think_ this is now ready to go. There's some missing documentation, but also this will allow us to delete most of `utils.py`.  Note that `compare` and summary could now work with pystan objects, though I took out some of the arguments to make them simpler and maybe less flexible. The summary also does not look great, but it is computationally fine - I think this should go in, and we add an issue to improve pretty-printing.

image

image

File: arviz/stats/stats.py [hunk 1]
```diff
@@ -221,8 +237,9 @@ def _ic_matrix(ics, ic_i):
     for idx, val in enumerate(ics.index):
         ic = ics.loc[val][ic_i]
         if len(ic) != rows:
-            raise ValueError('The number of observations should be the same '
-                             'across all models')
+            raise ValueError(
+                "The number of observations should be the same " "across all models"
```
Comment (canyon289) [inline arviz/stats/stats.py]: Not picky, redundant parentheses here?

Comment (ColCarroll) [reply arviz/stats/stats.py]: Fixed - I actually ran `black` on the file as an experiment - this might have been part of the result.

**Comment by ahartikainen** [comment]
LGTM

File: arviz/stats/diagnostics.py [hunk 2]
```diff
@@ -1,29 +1,28 @@
...
-
-from ..utils import trace_to_dataframe, get_varnames
+from ..utils.xarray_utils import convert_to_dataset
 
 
 __all__ = ['effective_n', 'gelman_rubin', 'geweke']
 
 
-def effective_n(trace, varnames=None, round_to=2):
+def effective_n(data, *, var_names=None):
```
Comment (junpenglao) [inline arviz/stats/diagnostics.py]: I didnt know you can use `*` as input - what does it do?

Comment (canyon289) [reply arviz/stats/diagnostics.py]: It prevents people from using positional arguments

**Comment by canyon289** [comment]
LGTM as well

**Pull request merged**